File: Runtime/Rust/src/rpc/router/calls.rs
use std::future::Future;
use std::marker::PhantomData;
use std::num::NonZeroU16;
use std::pin::Pin;
use std::sync::Weak;
use std::task::{Context, Poll};
use std::time::{Duration, Instant};

use tokio::sync::oneshot;

use crate::rpc::datagram::RpcResponseHeader;
use crate::rpc::error::{RemoteRpcResponse, TransportError, TransportResult};
use crate::rpc::{
    Datagram, DatagramInfo, Deadline, LocalRpcError, LocalRpcResponse, RouterContext,
};
use crate::{OwnedRecord, Record, SliceWrapper};

pub struct TypedRequestHandle<'r, R>
    where R: Record<'r>
{
    _ret_type: PhantomData<&'r R>,
    inner: RequestHandle,
}

/// Request handle to allow sending your response to the remote.
pub struct RequestHandle {
    /// Weak reference to the context we will need to send datagrams.
    ctx: Weak<RouterContext>,
    details: InnerCallDetails,
}

impl<'r, R> From<RequestHandle> for TypedRequestHandle<'r, R>
    where R: Record<'r>
{
    fn from(inner: RequestHandle) -> Self {
        Self {
            inner,
            _ret_type: PhantomData::default(),
        }
    }
}

impl<'r, R> TypedRequestHandle<'r, R>
    where R: Record<'r>
{
    pub async fn send_response(
        self,
        response: LocalRpcResponse<&R>,
    ) -> TransportResult {
        match response {
            Ok(record) => self.send_ok_response(record).await,
            Err(LocalRpcError::DeadlineExceded) => {
                // do nothing, no response needed as the remote should forget automatically.
                Ok(())
            }
            Err(LocalRpcError::CustomError(code, msg)) => {
                self.inner.send_error_response(code, Some(&msg)).await
            }
            Err(LocalRpcError::CustomErrorStatic(code, msg)) => {
                let msg = if msg.is_empty() { None } else { Some(msg) };
                self.inner.send_error_response(code, msg).await
            }
            Err(LocalRpcError::NotSupported) => self.inner.send_call_not_supported_response().await,
        }
    }

    /// Send a response to a call.
    pub async fn send_ok_response<'a, 'b: 'a>(
        self,
        record: &R,
    ) -> TransportResult {
        self.inner.send_ok_response_raw(
            &record
                .serialize_to_vec()
                .expect("Attempted to serialize an invalid response record"),
        )
            .await
    }
}

impl RequestHandle {
    pub(crate) fn new(ctx: Weak<RouterContext>, datagram: &Datagram) -> Self {
        debug_assert!(datagram.is_request(), "Must be a request");
        Self {
            ctx,
            details: InnerCallDetails {
                timeout: datagram.timeout(),
                since: Instant::now(),
                call_id: datagram
                    .call_id()
                    .expect("Request datagrams must have a valid call ID"),
            },
        }
    }

    pub async fn send_ok_response_raw(self, data: &[u8]) -> TransportResult {
        if let Some(ctx) = self.ctx.upgrade() {
            ctx.send(&Datagram::RpcResponseOk {
                header: RpcResponseHeader {
                    id: self.call_id().into(),
                },
                data: SliceWrapper::Cooked(data),
            })
            .await
        } else {
            Err(TransportError::NotConnected)
        }
    }

    pub async fn send_error_response(self, code: u32, msg: Option<&str>) -> TransportResult {
        if let Some(ctx) = self.ctx.upgrade() {
            ctx.send(&Datagram::RpcResponseErr {
                header: RpcResponseHeader {
                    id: self.call_id().into(),
                },
                code,
                info: msg.unwrap_or(""),
            })
            .await
        } else {
            Err(TransportError::NotConnected)
        }
    }

    pub async fn send_unknown_call_response(self) -> TransportResult {
        if let Some(ctx) = self.ctx.upgrade() {
            ctx.send(&Datagram::RpcResponseUnknownCall {
                header: RpcResponseHeader {
                    id: self.call_id().into(),
                },
            })
            .await
        } else {
            Err(TransportError::NotConnected)
        }
    }

    pub async fn send_invalid_sig_response(self, expected_sig: u32) -> TransportResult {
        if let Some(ctx) = self.ctx.upgrade() {
            ctx.send(&Datagram::RpcResponseInvalidSignature {
                header: RpcResponseHeader {
                    id: self.call_id().into(),
                },
                signature: expected_sig,
            })
            .await
        } else {
            Err(TransportError::NotConnected)
        }
    }

    pub async fn send_call_not_supported_response(self) -> TransportResult {
        if let Some(ctx) = self.ctx.upgrade() {
            ctx.send(&Datagram::RpcResponseCallNotSupported {
                header: RpcResponseHeader {
                    id: self.call_id().into(),
                },
            })
            .await
        } else {
            Err(TransportError::NotConnected)
        }
    }

    pub async fn send_decode_error_response(self, info: Option<&str>) -> TransportResult {
        if let Some(ctx) = self.ctx.upgrade() {
            ctx.send_decode_error_response(Some(self.call_id()), info)
                .await
        } else {
            Err(TransportError::NotConnected)
        }
    }
}

impl<'r, R> AsRef<InnerCallDetails> for TypedRequestHandle<'r, R>
    where R: Record<'r>
{
    fn as_ref(&self) -> &InnerCallDetails {
        &self.inner.details
    }
}

impl AsRef<InnerCallDetails> for RequestHandle {
    fn as_ref(&self) -> &InnerCallDetails {
        &self.details
    }
}

pub(super) fn new_pending_response<R>(
    call_id: NonZeroU16,
    timeout: Option<Duration>,
) -> (Box<dyn ResponseHandle>, PendingResponse<R>)
where
    R: 'static + OwnedRecord,
{
    let (tx, rx) = oneshot::channel::<RemoteRpcResponse<R>>();
    let details = InnerCallDetails {
        timeout,
        since: Instant::now(),
        call_id,
    };
    (
        Box::new(ResponseHandleImpl {
            tx: Some(tx),
            details,
        }),
        PendingResponse { rx, details },
    )
}

/// A pending response handle which resolves into the response from the remote. Use the trait for
/// type erasure since we can't store a bunch of different T types in one table.
///
/// Would just send `&[u8]` but doing that over a channel is not possible so we have to deserialize
/// an owned copy and _then_ send.
struct ResponseHandleImpl<T> {
    tx: Option<oneshot::Sender<RemoteRpcResponse<T>>>,
    details: InnerCallDetails,
}

impl<T> AsRef<InnerCallDetails> for ResponseHandleImpl<T> {
    fn as_ref(&self) -> &InnerCallDetails {
        &self.details
    }
}

pub trait ResponseHandle: Send + Sync + CallDetails {
    fn resolve(&mut self, value: RemoteRpcResponse<&[u8]>);
}

impl<R: OwnedRecord + Send + Sync> ResponseHandle for ResponseHandleImpl<R> {
    fn resolve(&mut self, value: RemoteRpcResponse<&[u8]>) {
        let res = value.and_then(|v| Ok(R::deserialize(v)?));
        if let Some(tx) = self.tx.take() {
            if tx.send(res).is_err() {
                // TODO: log this? Receiver stopped listening.
            }
        }
    }
}

/// A pending response from the remote which will resolve once you receive their reply.
pub(super) struct PendingResponse<T> {
    rx: oneshot::Receiver<RemoteRpcResponse<T>>,
    details: InnerCallDetails,
}

impl<T> AsRef<InnerCallDetails> for PendingResponse<T> {
    fn as_ref(&self) -> &InnerCallDetails {
        &self.details
    }
}

impl<T> Future for PendingResponse<T> {
    type Output = RemoteRpcResponse<T>;

    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        match Pin::new(&mut self.rx).poll(cx) {
            Poll::Ready(Ok(res)) => Poll::Ready(res),
            Poll::Ready(Err(..)) if self.is_expired() => {
                Poll::Ready(Err(TransportError::Timeout.into()))
            }
            Poll::Ready(Err(..)) => Poll::Ready(Err(TransportError::CallDropped.into())),
            // timeouts are handled by the call_table and not the pending response itself
            Poll::Pending => Poll::Pending,
        }
    }
}

#[derive(Copy, Clone, Debug)]
pub struct InnerCallDetails {
    /// How long this call is allowed to be pending for. If None, no timeout is specified.
    ///
    /// Warning: No timeout will lead to memory leaks if the transport does not notify the router
    /// of dropped/missing data.
    timeout: Option<Duration>,

    /// The instant at which this call was sent/received.
    since: Instant,

    /// The unique call ID assigned by us, the caller.
    call_id: NonZeroU16,
}

pub trait CallDetails {
    fn call_id(&self) -> NonZeroU16;

    fn duration(&self) -> Duration {
        Instant::now() - self.since()
    }

    fn since(&self) -> Instant;

    fn timeout(&self) -> Option<Duration>;

    fn is_expired(&self) -> bool {
        if let Some(timeout) = self.timeout() {
            self.duration() >= timeout
        } else {
            false
        }
    }

    fn expires_at(&self) -> Option<Instant> {
        self.timeout().map(|t| self.since() + t)
    }

    fn deadline(&self) -> Deadline {
        self.expires_at().into()
    }
}

impl CallDetails for InnerCallDetails {
    fn call_id(&self) -> NonZeroU16 {
        self.call_id
    }

    fn since(&self) -> Instant {
        self.since
    }

    fn timeout(&self) -> Option<Duration> {
        self.timeout
    }
}

impl<T: AsRef<InnerCallDetails>> CallDetails for T {
    fn call_id(&self) -> NonZeroU16 {
        self.as_ref().call_id()
    }

    fn since(&self) -> Instant {
        self.as_ref().since()
    }

    fn timeout(&self) -> Option<Duration> {
        self.as_ref().timeout()
    }
}

#[cfg(test)]
mod test {
    use std::num::NonZeroU16;
    use std::time::Duration;

    use crate::rpc::calls::new_pending_response;
    use crate::rpc::test_struct::TestStruct;
    use crate::rpc::{CallDetails, RemoteRpcError, TransportError};
    use crate::timeout;

    #[tokio::test]
    async fn resolves() {
        // happy path
        let (mut handle1, pending1) =
            new_pending_response::<TestStruct>(1.try_into().unwrap(), None);
        let (mut handle2, pending2) =
            new_pending_response::<TestStruct>(2.try_into().unwrap(), timeout!(10 s));

        let data1 = [65];
        let data2 = [223];

        assert_eq!(pending1.call_id(), NonZeroU16::new(1).unwrap());
        assert_eq!(pending1.timeout(), None);
        assert_eq!(pending2.timeout(), Some(Duration::from_secs(10)));
        assert!(!pending1.is_expired());
        assert!(!pending2.is_expired());

        handle1.resolve(Ok(&data1));
        handle2.resolve(Ok(&data2));

        assert_eq!(pending1.await.unwrap(), TestStruct { v: data1[0] });
        assert_eq!(pending2.await.unwrap(), TestStruct { v: data2[0] });
    }

    #[tokio::test]
    async fn resolves_err() {
        // when there is a transport error
        let (mut handle, pending) = new_pending_response::<TestStruct>(1.try_into().unwrap(), None);
        handle.resolve(Err(RemoteRpcError::TransportError(
            TransportError::CallDropped,
        )));
        assert!(matches!(
            pending.await,
            Err(RemoteRpcError::TransportError(TransportError::CallDropped))
        ));
    }

    #[tokio::test]
    async fn resolves_timeout() {
        // when the handle gets dropped due to a timeout
        let (handle, pending) = new_pending_response::<TestStruct>(
            1.try_into().unwrap(),
            Some(Duration::from_millis(10)),
        );
        assert!(!handle.is_expired());
        assert!(!pending.is_expired());

        tokio::time::sleep(Duration::from_millis(10)).await;
        assert!(handle.is_expired());
        assert!(pending.is_expired());
        drop(handle);
        let r = pending.await;
        assert!(
            matches!(
                r,
                Err(RemoteRpcError::TransportError(TransportError::Timeout))
            ),
            "{r:?}"
        );
    }

    #[tokio::test]
    async fn resolves_dropped() {
        // when the handle gets dropped but it has not timed out
        let (_, pending) = new_pending_response::<TestStruct>(1.try_into().unwrap(), None);
        assert!(matches!(
            pending.await,
            Err(RemoteRpcError::TransportError(TransportError::CallDropped))
        ));
    }
}

File: Runtime/TypeScript/src/rpc/router/calls.ts
import { RouterContext } from "./context";
import { Deadline } from "../deadlines";
import {
  assertUnreachable,
  BebopRecordImpl,
  BebopRuntimeError,
} from "../../index";
import {
  LocalRpcError,
  LocalRpcErrorVariants,
  TransportError,
  TransportErrorDeserializationError,
  TransportErrorVariants,
} from "../error";
import { isNativeError, isUint8Array } from "util/types";
import { RecordTypeOf } from "../../record";
import {
  RpcDecodeError,
  RpcResponseCallNotSupported,
  RpcResponseErr,
  RpcResponseInvalidSignature,
  RpcResponseOk,
  RpcResponseUnknownCall,
} from "../../generated/datagram";
import { IDatagram } from "../index";
import DatagramInfo from "../datagram-info";
import * as assert from "assert";

/** Request handle to allow sending your response to the remote. */
export class RequestHandle implements CallDetails {
  private responseSent = false;
  readonly callDetails: CallDetailsInner;
  private readonly ctx: WeakRef<RouterContext>;

  /** @ignore */
  constructor(ctx: WeakRef<RouterContext>, dgram: IDatagram);
  /** @ignore */
  constructor(rh: RequestHandle);
  constructor(...args: unknown[]) {
    const targs = args as
      | [rh: RequestHandle]
      | [ctx: WeakRef<RouterContext>, dgram: IDatagram];
    if (targs[0] instanceof RequestHandle) {
      const [rh] = targs;
      this.responseSent = rh.responseSent;
      this.callDetails = rh.callDetails;
      this.ctx = rh.ctx;
    } else {
      const [ctx, dgram] = [targs[0], targs[1]!];
      this.ctx = ctx;
      this.callDetails = {
        timeout: DatagramInfo.timeoutS(dgram),
        since: Date.now(),
        callId: DatagramInfo.callId(dgram),
      };
    }
  }

  async sendOkResponseRaw(record: Uint8Array): Promise<void> {
    const ctx = this.$ctx;
    this.verifyNotAlreadyResponded();
    await ctx.send({
      discriminator: RpcResponseOk.discriminator,
      value: {
        header: {
          id: CallDetails.callId(this),
        },
        data: record,
      },
    });
  }

  async sendErrorResponse(code: number, msg?: string): Promise<void> {
    assert(
      code >= 0 && code <= 0xffffffff && Number.isInteger(code),
      "Invalid error code provided. Must be a valid uint32."
    );
    this.verifyNotAlreadyResponded();
    await this.$ctx.send({
      discriminator: RpcResponseErr.discriminator,
      value: {
        header: {
          id: CallDetails.callId(this),
        },
        code,
        info: msg ?? "",
      },
    });
  }

  async sendUnknownCallResponse(): Promise<void> {
    this.verifyNotAlreadyResponded();
    await this.$ctx.send({
      discriminator: RpcResponseUnknownCall.discriminator,
      value: {
        header: { id: CallDetails.callId(this) },
      },
    });
  }

  async sendInvalidSigResponse(expected_sig: number): Promise<void> {
    assert(
      expected_sig >= 0 &&
        expected_sig <= 0xffffffff &&
        Number.isInteger(expected_sig),
      "Invalid expected signature provided. Must be a valid uint32."
    );
    this.verifyNotAlreadyResponded();
    await this.$ctx.send({
      discriminator: RpcResponseInvalidSignature.discriminator,
      value: {
        header: { id: CallDetails.callId(this) },
        signature: expected_sig,
      },
    });
  }

  async sendCallNotSupportedResponse(): Promise<void> {
    this.verifyNotAlreadyResponded();
    await this.$ctx.send({
      discriminator: RpcResponseCallNotSupported.discriminator,
      value: {
        header: {
          id: CallDetails.callId(this),
        },
      },
    });
  }

  async sendDecodeErrorResponse(
    info: string | TransportErrorDeserializationError = ""
  ): Promise<void> {
    this.verifyNotAlreadyResponded();
    await this.$ctx.send({
      discriminator: RpcDecodeError.discriminator,
      value: {
        header: {
          id: CallDetails.callId(this),
        },
        info: typeof info == "string" ? info : info.error.message,
      },
    });
  }

  /** Get a strong reference to the router context and throw a not connected error if it fails */
  private get $ctx(): RouterContext {
    const ctx = this.ctx.deref();
    if (!ctx) {
      throw new TransportError({
        discriminator: TransportErrorVariants.NotConnected,
      });
    }
    return ctx;
  }

  private verifyNotAlreadyResponded(): void {
    if (this.responseSent) {
      throw new TransportError({
        discriminator: TransportErrorVariants.ResponseAlreadySent,
      });
    }
    this.responseSent = true;
  }
}

export class TypedRequestHandle<
  RI extends BebopRecordImpl
> extends RequestHandle {
  constructor(private readonly recordImpl: RI, inner: RequestHandle) {
    super(inner);
  }

  async sendResponse(
    response: RecordTypeOf<RI> | Uint8Array | Error
  ): Promise<void> {
    if (response instanceof LocalRpcError) {
      const e = response.inner;
      switch (e.discriminator) {
        case LocalRpcErrorVariants.DeadlineExceeded:
          // do nothing, no response needed as the remote should forget automatically.
          break;
        case LocalRpcErrorVariants.Custom:
          await this.sendErrorResponse(e.code, e.info);
          break;
        case LocalRpcErrorVariants.NotSupported:
          await this.sendCallNotSupportedResponse();
          break;
        default:
          assertUnreachable(e);
      }
    } else if (isNativeError(response)) {
      console.error("Unhandled exception: ", response);
      await this.sendErrorResponse(
        0xffffffff,
        `Unhandled internal exception: ${response.message}`
      );
    } else {
      await this.sendOkResponse(response);
    }
  }

  async sendOkResponse(record: RecordTypeOf<RI> | Uint8Array): Promise<void> {
    // TODO: Don't clone if https://github.com/RainwayApp/bebop/issues/209 removes the pooled view
    const raw = isUint8Array(record)
      ? record
      : new Uint8Array(this.recordImpl.encode(record));
    return this.sendOkResponseRaw(raw);
  }
}

/** @ignore */
export interface ResponseHandle extends CallDetails {
  resolve(data: Uint8Array): void;

  reject(err: BebopRuntimeError): void;
}

/** @ignore */
export interface PendingResponse<TRecord>
  extends CallDetails,
    PromiseLike<TRecord> {}

/**
 * @ignore
 *
 * @param recordImpl The record implementation which we need for deserializing.
 * @param callId Valid, user-assigned call id.
 * @param timeout Optional timeout in seconds.
 */
export function newPendingResponse<RI extends BebopRecordImpl>(
  recordImpl: RI,
  callId: number,
  timeout?: number
): [tx: ResponseHandle, rx: PendingResponse<RecordTypeOf<RI>>] {
  let resolve: ResponseHandle["resolve"];
  let reject: ResponseHandle["reject"];

  const promise = new Promise<Uint8Array>((res, rej) => {
    resolve = res;
    reject = rej;
  }).then((buf) => recordImpl.decode(buf) as RecordTypeOf<RI>);

  const callDetails: CallDetails["callDetails"] = Object.freeze({
    callId,
    since: Date.now(),
    timeout,
  });

  return [
    {
      resolve: resolve!,
      reject: reject!,
      callDetails,
    },
    {
      then: promise.then.bind(promise),
      callDetails,
    },
  ];
}

interface CallDetailsInner {
  /**
   * How long this call is allowed to be pending for (in seconds). If undefined, no timeout is specified.
   *
   * Warning: No timeout will lead to memory leaks if the transport does not notify the router
   * of dropped/missing data.
   */
  readonly timeout?: number;
  /** The instant at which this call was sent/received. (JS Epoch in ms) */
  readonly since: number;
  /** The unique call ID assigned by us, the caller. */
  readonly callId: number;
}

interface CallDetails {
  readonly callDetails: CallDetailsInner;
}

export const CallDetails = {
  callId(cd: CallDetails): number {
    return cd.callDetails.callId;
  },
  /** JS Epoch of when this was created. */
  since(cd: CallDetails): number {
    return cd.callDetails.since;
  },
  durationMs(cd: CallDetails): number {
    return Date.now() - this.since(cd);
  },
  durationS(cd: CallDetails): number {
    return this.durationMs(cd) / 1000;
  },
  timeoutS(cd: CallDetails): number | undefined {
    return cd.callDetails.timeout === undefined
      ? undefined
      : cd.callDetails.timeout;
  },
  timeoutMs(cd: CallDetails): number | undefined {
    return cd.callDetails.timeout === undefined
      ? undefined
      : cd.callDetails.timeout * 1000;
  },
  isExpired(cd: CallDetails): boolean {
    const timeout = this.timeoutMs(cd);
    return timeout !== undefined && this.durationMs(cd) >= timeout;
  },
  /** JS Epoch at which this expires */
  expiresAt(cd: CallDetails): number | undefined {
    const timeout = this.timeoutMs(cd);
    return timeout === undefined ? undefined : this.since(cd);
  },
  deadline(cd: CallDetails): Deadline {
    return new Deadline(this.expiresAt(cd));
  },
} as const;
